src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/RepoFetchingSkyKeyComputeState.java

```diff
@@ -79,17 +112,141 @@ enum Signal {
    */
   final Map<RepoRecordedInput, String> recordedInputValues = new TreeMap<>();
 
+  RepoFetchingSkyKeyComputeState(String repositoryName,
+      Function<RepoFetchingSkyKeyComputeState, Runnable> threadFunc) {
+    this.repositoryName = repositoryName;
+    workerThread =
+        Thread.ofVirtual()
+            .name("starlark-repository-" + repositoryName)
+            .start(threadFunc.apply(this));
+  }
+
   SkyFunction.Environment signalForFreshEnv() throws InterruptedException {
-    signalQueue.put(Signal.RESTART);
-    return delegateEnvQueue.take();
+    Response response = sendRequest(new Request.NewDependencies());
+    if (!(response instanceof Response.Restarting)) {
+      throw new IllegalStateException("Unexpected response %s" + response);
+    }
+
+    response = sendRequest(new Request.GetEnvironment());
+    if (response instanceof Response.Environment environmentResponse) {
+      return environmentResponse.environment;
+    } else {
+      throw new IllegalStateException("Unexpected response %s" + response);
+    }
+  }
+
+  private void logMaybe(String msg) {
+    // System.err.println(msg);
   }
 
+  Response sendRequest(Request request) throws InterruptedException {
+    int seq = nextRequestSequenceNo++;
+    Packet<Request> packet = new Packet(seq, request);
+    logMaybe(String.format("LOG %s/worker: sending request %s", repositoryName, packet));
+    requestQueue.put(packet);
+    while (true) {
+      Packet<Response> response = responseQueue.take();
+      if (response.sequenceNo == seq) {
```

Comment:
Under what circumstances will we get out of order responses? Can we just assert on the seq?

(same question for `sendRequestUninterruptibly`).

Reply:
Hehe. This is complicated :) What happen is this:

1. The worker thread sends a request
2. The worker thread gets interrupted. Sends another `Failure` request
3. The host thread gets around to reply to the message in (1)
4. The worker thread receives the answer for (3) while waiting for the response in (1)

I'll add a comment to this effect (once I figured out the rest)